src/transaction/log_recovery_context.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2008 Search Solution Corporation
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include "log_recovery_context.hpp"
+
+#include "disk_manager.h"
+#include "file_io.h"
+#include "log_impl.h"
+#include "server_type.hpp"
+#include "thread_manager.hpp"
+
+bool
+get_disk_checkpoint_min_lsa (THREAD_ENTRY *thread_p, VOLID volid, void *lsa)
+{
+  // the function signature is adapted to fileio_map_mounted. that last argument is actually a pointer to log_lsa.
+  assert (lsa != nullptr);
+  log_lsa &global_checkpoint_lsa = * (log_lsa *) lsa;
+  log_lsa local_checkpoint_lsa = NULL_LSA;
+
+  const int ret = disk_get_checkpoint (thread_p, volid, &local_checkpoint_lsa);
+  if (ret == NO_ERROR)
+    {
+      if (global_checkpoint_lsa.is_null () || local_checkpoint_lsa < global_checkpoint_lsa)
+	{
+	  global_checkpoint_lsa = local_checkpoint_lsa;
+	}
+    }
+
+  return true;
+}
+
+log_recovery_context::log_recovery_context ()
+{
+  m_is_page_server = get_server_type () == SERVER_TYPE_PAGE;
+  m_checkpoint_lsa = log_Gl.hdr.chkpt_lsa;
+}
+
+void
+log_recovery_context::init_for_restore (const time_t *stopat_p)
+{
+  m_is_restore_from_backup = true;
+  if (stopat_p)
+    {
+      m_restore_stop_point = *stopat_p;
+    }
+
+  /* we may have to start from an older checkpoint... */
+  (void) fileio_map_mounted (&cubthread::get_entry (), get_disk_checkpoint_min_lsa, &m_checkpoint_lsa);
```

Comment:
I find it a bit odd to bury this aspect here from its prominent place in `log_recovery`. This class is supposed to provide some bookkeeping container for recovery and add as little business logic as possible.

Reply:
You make a valid point :). The logic of getting min LSA from disk was moved back to log_recovery().